open-craft/learner-profile-dashboard
# Display information about profile and section-level completion

Cf. [SE-952](https://tasks.opencraft.com/browse/SE-952)

This PR adds functionality for displaying profile and section-level completion (in the form of percentages) on full-LPD embeds. It also adds logic for automatically updating completion information when a learner successfully submits answers belonging to a specific section of an LPD.

### Definition of completeness

#### Profile

A learner’s profile is *complete* if and only if they submitted an answer to each available question. What constitutes an answer differs based on question type (cf. below).

A learner’s profile is *partially complete* if they submitted answers for less than 100% of all questions belonging to a given LPD instance.

If a profile does not have any sections associated with it, it is considered 0% complete by default.

#### Section

Similarly, a section is complete if and only if a learner submitted an answer to each question belonging to that section. It is partially complete if the learner submitted answers for less than 100% of all questions belonging to it.

If a section does not have any questions associated with it, it is considered 0% complete by default.

### Screenshots

Profile and section-level completeness:

![lpd-completion-percentages](https://user-images.githubusercontent.com/961441/56671406-e09c5900-66b4-11e9-9d13-6a00141c0c44.png)

### Test instructions

1. If you haven't already, create an account on [HGSE stage](https://hgse-stage.opencraft.hosting/) and enroll in the [demo course](https://hgse-stage.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/about).

1. Navigate to [this unit](https://hgse-stage.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/jump_to/block-v1:edX+DemoX+Demo_Course+type@vertical+block@45c7cedb4bfe46f4a68c78787151cfb5).

1. For each section:

    1. Answer some questions and click "Submit your answers".

    1. Verify that the completion percentage for the section as well as the completion percentage for the LPD as a whole (displayed at the very top) update to reflect the changes that you made.

        - Note that if you only *change* an existing answer and click "Submit your answers" without providing any *additional* answers, the completion percentages shouldn't change.

        - If you're not sure about a specific update to profile and section-level completion percentages, have a look at relevant information about the LPD's criteria for considering a question answered. These criteria differ based on question type:

            - [Qualitative questions](https://github.com/open-craft/learner-profile-dashboard/blob/profile-completion/lpd/models.py#L277)
            - [Multiple choice questions](https://github.com/open-craft/learner-profile-dashboard/blob/profile-completion/lpd/models.py#L439-L440)
            - [Ranking questions](https://github.com/open-craft/learner-profile-dashboard/blob/profile-completion/lpd/models.py#L525-L537)
            - [Likert scale questions](https://github.com/open-craft/learner-profile-dashboard/blob/profile-completion/lpd/models.py#L595-L596)

    1. Reload the page.

    1. Verify that the changes to completion percentages that you observed in the previous step persist.

    1. Answer some additional questions and reload the page *without* clicking "Submit your answers".

    1. Verify that completion percentages reset to values that you observed previously.

### Author notes

In terms of UX, an alternative to updating completion information after successfully submitting answers via the "Submit your answers" button would be to update completion information as a learner moves from question to question, adding/updating their answers along the way. However, this is not an option for the current LPD implementation: Answer data is recorded/updated in the DB only after a learner explicitly indicates that they want to submit their answers for a section (by clicking the corresponding "Submit your answers" button). (In other words, no changes are made to answer data stored in the DB as the learner moves from question to question.) If they navigate away without clicking "Submit your answers", the current set of answers from the learner that is stored in the DB doesn't change, and neither do the learner's completion statistics for individual sections and the LPD as a whole. So updating completion information as the learner moves from question to question wouldn't make sense -- it wouldn't accurately reflect the current state of the DB, and could potentially confuse the learner since the changes to their completion statistics wouldn't necessarily be permanent. (They'd reset to previous values if the learner came back to their learner profile after forgetting to submit pending changes to their answers.)

### Reviewers
